torch/csrc/jit/codegen/cuda/scheduler/registry.cpp

```diff
@@ -748,135 +778,111 @@ class NormalizationScheduler : public SchedulerEntry {
+    for (size_t it = 1; it < reduction_tvs.size(); it++) {
+      if (!checkEquivalence(
+              reduction_tvs[it - 1], reduction_tvs[it], root_map)) {
+        return false;
       }
     }
-    auto& persistent_buffers = persistent_buffer_data.read();
+
+    return true;
+  }
+
+  static bool canScheduleRunTime(
+      Fusion* fusion,
+      SchedulerRuntimeInfo& runtime_info,
+      HeuristicSummary* data_cache = nullptr) {
+    FUSER_PERF_SCOPE("NormalizationScheduler::canSchedule");
+
+    auto reduction_tv_entry =
+        HeuristicSummaryEntry<HeuristicCompileTime::ReductionTVs>(
+            data_cache, [&fusion]() {
+              return std::make_unique<std::vector<TensorView*>>(
+                  scheduler_utils::getReductionTvs(fusion));
+            });
+
+    auto& reduction_tvs = reduction_tv_entry.get();
+
+    auto persistent_buffer_info_entry =
+        HeuristicSummaryEntry<HeuristicCompileTime::PersistentBufferInfo>(
+            data_cache, [&fusion]() {
+              return std::make_unique<scheduler_utils::PersistentBufferInfo>(
+                  scheduler_utils::persistentBuffers(fusion));
+            });
+
+    auto& persistent_buffers = persistent_buffer_info_entry.get();
 
     auto persistent_buffer_size = scheduler_utils::persistentBufferSize(
         fusion, runtime_info, persistent_buffers, data_cache);
     if (persistent_buffer_size * 4 > scheduler_utils::register_file_size * 3) {
       return false;
     }
 
-    // TODO: really need to make inserting an entry into data_cache easier to do
-    HeuristicCacheAccessor<bool> has_post_reduction_bcast_data;
-
-    if (data_cache && !data_cache->isRecording()) {
-      has_post_reduction_bcast_data.writeTemporary(
-          data_cache->getHasPostReductionBCast());
-    } else {
-      has_post_reduction_bcast_data.writeNew(
-          SchedulerTopologyChecker::hasPostReductionBCast(fusion));
-      if (data_cache && data_cache->isRecording()) {
-        data_cache->setHasPostReductionBCast(
-            has_post_reduction_bcast_data.read());
-      }
-    }
+    auto reduction_topology_info_entry = HeuristicSummaryEntry<
+        HeuristicCompileTime::ReductionTopologyInfo>(
+        data_cache, [&fusion, &reduction_tvs]() {
+          HeuristicCompileTime::ReductionTopologyCheck topology_check_data;
 
-    HeuristicCacheAccessor<bool> supported_post_reduction_fusion_data;
-
-    if (data_cache && !data_cache->isRecording()) {
-      supported_post_reduction_fusion_data.writeTemporary(
-          data_cache->getSupportedPostReductionFusion());
-    } else {
-      supported_post_reduction_fusion_data.writeNew(
-          SchedulerTopologyChecker::supportedPostReductionFusion(
-              fusion, reduction_tvs));
-      if (data_cache && data_cache->isRecording()) {
-        data_cache->setSupportedPostReductionFusion(
-            supported_post_reduction_fusion_data.read());
-      }
-    }
+          topology_check_data.has_post_reduction_bcast =
```

Comment:
I think these two topology checks can go in compile time check.

Reply:
The topology checks are used in a runtime info dependent way in `canSchedule`. Their results were only used when `persistent buffer size <=1`. So they need to be cached. 

I was trying to avoid the complication of exposing data cache in compile time check because data cache may contain runtime information.